p2p_beef_tx.go

```diff
@@ -39,6 +40,151 @@ type DecodedBEEF struct {
+	var merkleRoots []string
+	for _, cmp := range dBeef.CMPSlice {
+		partialMerkleRoots, err := cmp.CalculateMerkleRoots()
+		if err != nil {
+			return nil, err
+		}
+		merkleRoots = append(merkleRoots, partialMerkleRoots...)
+	}
+	return merkleRoots, nil
+}
+
+// ExecuteSimplifiedPaymentVerification executes the SPV for decoded BEEF tx
+func (dBeef *DecodedBEEF) ExecuteSimplifiedPaymentVerification() error {
+	if len(dBeef.ProcessedTxData.Transaction.Inputs) == 0 {
+		return errors.New("invalid input, no inputs")
+	}
+
+	if len(dBeef.ProcessedTxData.Transaction.Outputs) == 0 {
+		return errors.New("invalid output, no outputs")
+	}
+
+	// check locktime and sequence
+	if dBeef.ProcessedTxData.Transaction.LockTime == 0 {
+		for _, input := range dBeef.ProcessedTxData.Transaction.Inputs {
+			if input.SequenceNumber != 0xffffffff {
+				return errors.New("invalid sequence")
+			}
+		}
+	} else {
+		return errors.New("invalid locktime or sequence")
+	}
+
+	// check inputs and outputs
+	inputSum, outputSum := uint64(0), uint64(0)
+	for i, input := range dBeef.ProcessedTxData.Transaction.Inputs {
+		input2 := dBeef.InputsTxData[i]
+		inputSum += input2.Transaction.Outputs[input.PreviousTxOutIndex].Satoshis
+	}
+	for _, output := range dBeef.ProcessedTxData.Transaction.Outputs {
+		outputSum += output.Satoshis
+	}
+
+	// Check if the output sum is not higher than the input sum
+	if inputSum <= outputSum {
+		return errors.New("invalid input and output sum, outputs can not be larger than inputs")
+	}
+
+	// Verify scripts
+	for _, input := range dBeef.ProcessedTxData.Transaction.Inputs {
+		txId := input.PreviousTxID()
+		for j, input2 := range dBeef.InputsTxData {
+			if input2.Transaction.TxID() == string(txId) {
+				result := verifyScripts(dBeef.ProcessedTxData.Transaction, input2.Transaction, j)
+				if !result {
+					return errors.New("invalid script")
+				}
+				break
+			}
+		}
+	}
+
+	return nil
+}
+
+func (cmp *CompoundMerklePath) CalculateMerkleRoots() ([]string, error) {
+	merkleRoots := make([]string, 0)
+	cmpCopy := *cmp
+
+	// Iterate through first layer
+	for tx, offset := range cmpCopy[len(cmpCopy)-1] {
+		// Calculate merkle root for one tx
+		merkleRoot, err := calculateMerkleRoot(tx, offset, cmpCopy)
+		if err != nil {
+			return nil, err
+		}
+		merkleRoots = append(merkleRoots, merkleRoot)
+
+	}
+	return merkleRoots, nil
+}
+
+// Verify locking and unlocking scripts pair
+func verifyScripts(tx, prevTx *bt.Tx, inputIdx int) bool {
+	input := tx.InputIdx(inputIdx)
+	prevOutput := prevTx.OutputIdx(int(input.PreviousTxOutIndex))
+
+	if err := interpreter.NewEngine().Execute(
+		interpreter.WithTx(tx, inputIdx, prevOutput),
+		interpreter.WithForkID(),
+		interpreter.WithAfterGenesis(),
+	); err != nil {
+		fmt.Println(err)
+		return false
+	}
+	return true
+}
+
+func calculateMerkleRoot(baseTx string, offset uint64, cmp []map[string]uint64) (string, error) {
+	// Iterate through layers
+	for i := len(cmp) - 1; i >= 0; i-- {
```

Comment:
@pawellewandowski98 this comment is not related to this PR,
But actually to BEEF decoding.
In current implementation after decoding we get the CMP that is in different order then the one before encoding. In my opinion we should change that, because:
1. it is misleading - I would assume that when I encode something then I will receive the same thing after decoding, what is not true here.
2. it is cumbersome to reason about the code that is processing this decode CMP because it is in different order then the one that is described in documents (BRC) and the one before encoding.

Reply:
I agree, we should adjust it to BRC

Reply:
Created bug in JIRA
https://bitcoinassociation.atlassian.net/browse/BUX-272